prime-router/src/main/kotlin/fhirengine/engine/LookupTableValueSet.kt

```diff
@@ -0,0 +1,69 @@
+package gov.cdc.prime.router.fhirengine.engine
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import gov.cdc.prime.router.Metadata
+import gov.cdc.prime.router.fhirengine.translation.hl7.SchemaException
+import gov.cdc.prime.router.fhirengine.translation.hl7.ValueSetCollection
+import java.util.SortedMap
+
+/**
+ * Data class for configuring [LookupTableValueSet].
+ *
+ * [tableName]: name of the lookup table
+ *
+ * [keyColumn]: name of the lookup table column containing the key pair values
+ *
+ * [valueColumn]: name of the lookup table column containing the value pair values
+ */
+data class LookupTableValueSetConfig(
+    val tableName: String,
+    val keyColumn: String,
+    val valueColumn: String
+)
+
+/**
+ * Implementation of [ValueSetCollection] to allow valueSet to be retrieved from a lookup table.
+ * Provide [LookupTableValueSetConfig] to configure the lookup table source.
+ */
+class LookupTableValueSet
+(@JsonProperty("lookupTable") private val configData: LookupTableValueSetConfig) : ValueSetCollection {
+    private val metadata = Metadata.getInstance()
+    private var mapVal: SortedMap<String, String>? = null
+
+    override fun toSortedMap(): SortedMap<String, String> {
+        if (mapVal == null) {
+            buildMap()
+        }
+        return mapVal as SortedMap<String, String>
+    }
+
+    override fun getMappedValue(keyValue: String): String? {
+        val lowerSet = toSortedMap().mapKeys { it.key.lowercase() }
+        return lowerSet[keyValue.lowercase()]
+    }
+
+    override fun isNotEmpty(): Boolean {
+        return toSortedMap().isNotEmpty()
+    }
+
+    private fun buildMap() {
+        val lookupTable = metadata.findLookupTable(name = configData.tableName)
+            ?: throw SchemaException("Specified lookup table not found")
```

Comment:
Nit: This could potentially get moved into the constructor rather than waiting until this actually gets used